flaml/spark/utils.py

```diff
@@ -0,0 +1,173 @@
+    logger.debug("\ncheck Spark installation...This line should appear only once.\n")
+    if not _have_spark:
+        raise ImportError(
+            "use_spark=True requires installation of PySpark. "
+            "Please run pip install flaml[spark] and check "
+            "[here](https://spark.apache.org/docs/latest/api/python/getting_started/install.html) "
+            "for more details about installing Spark."
+        )
+
+    if _spark_major_minor_version[0] < 3:
+        raise ImportError("Spark version must be >= 3.0 to use flaml[spark]")
+
+    # Raise the original exception if SparkSession is not available
+    SparkSession.builder.getOrCreate()
+
+    return True
+
+
+def get_n_cpus(node="driver"):
+    """Get the number of CPU cores of the given type of node.
+
+    Args:
+        node: string | The type of node to get the number of cores. Can be 'driver' or 'executor'.
+            Default is 'driver'.
+
+    Returns:
+        An int of the number of CPU cores.
+    """
+    assert node in ["driver", "executor"]
+    try:
+        n_cpus = int(
+            SparkSession.builder.getOrCreate()
+            .sparkContext.getConf()
+            .get(f"spark.{node}.cores")
+        )
+    except (TypeError, RuntimeError):
+        n_cpus = os.cpu_count()
+    return n_cpus
+
+
+def with_parameters(trainable, **kwargs):
+    """Wrapper for trainables to pass arbitrary large data objects.
+
+    This wrapper function will store all passed parameters in the Spark
+    Broadcast variable.
+
+    Args:
+        trainable: Trainable to wrap.
+        **kwargs: parameters to store in object store.
+
+    Returns:
+        A new function with partial application of the given arguments
+        and keywords. The given arguments and keywords will be broadcasted
+        to all the executors.
+
+
+    ```python
+    import pyspark
+    import flaml
+    from sklearn.datasets import load_iris
+    def train(config, data=None):
+        if isinstance(data, pyspark.broadcast.Broadcast):
+            data = data.value
+        print(config, data)
+
+    data = load_iris()
+    with_parameters_train = flaml.spark.utils.with_parameters(train, data=data)
+    with_parameters_train(config=1)
+    train(config={"metric": "accuracy"})
+    ```
+    """
+
+    if not callable(trainable):
+        raise ValueError(
+            f"`with_parameters() only works with function trainables`. "
+            f"Got type: "
+            f"{type(trainable)}."
+        )
+
+    check_spark()
+    spark = SparkSession.builder.getOrCreate()
+
+    bc_kwargs = dict()
+    for k, v in kwargs.items():
+        bc_kwargs[k] = spark.sparkContext.broadcast(v)
+
+    return partial(trainable, **bc_kwargs)
+
+
+def customize_learner(learner_code="", file_name="mylearner"):
+    """Write customized learner/metric code contents to a file for importing.
+    It is necessary for using the customized learner in spark backend.
+    The path of the learner file will be returned.
```

Comment:
Don't really get the purpose of this. Why would we write code to a string and write it back out again?

Reply:
If a class is defined inline, it can't be passed to spark executors correctly, we need to import it from a file/module. This is a limitation of pyspark.